File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int nr = matrix.size();
        int nc = matrix[0].size();

        int first = 0;
        int last = nr * nc;
        while(first <= last) {
            int mid = first + (last - first) / 2;
            if(mid/nc >= nr || mid%nc >= nc) {
                last = mid - 1;
                continue;
            }
            if(matrix[mid/nc][mid%nc] == target) return true;
            else if(matrix[mid/nc][mid%nc] > target) last = mid - 1;
            else first = mid + 1;
        }

        return false;
    }
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int n= matrix.length;
        int m= matrix[0].length;
        int i=0;
        int j= m-1;
        while(i>=0 && j>=0 && i<n && j<m) {
            if(matrix[i][j] > target) {
                j--;
            } else if(matrix[i][j] < target) {
                i++;
            } else {
                return true;
            }
        }
        return false;
    }
}